### PR TITLE
Support for multi-sig schemes. ECC (K1/R1) and EdDSA are fully supported.

### PR DESCRIPTION
**Pull request by kchalkias**
1. Sign/verify/keygen operations are now scheme agnostic. There are no signWithEdDSA and other similar calls anymore, just sign().
2. Serialisation for EdDSA remained the same (as it is optimised), while the rest are using the default's java encoding (PKCS8 for private keys, X509 for public). Easy to change if we want to (but the serialised object will be slightly bigger due to the additional headers for PKCS8/X509 formats).
3. I2P recently implemented their own provider and I've changed to that for consistency reasons (the signature algorithm is now called NONEwithEdDSA and not SHA512withEdDSA to comply with the JCA Signature naming scheme - to clarify, there is still SHA512, but as it is an internal feature of the algorithm and we don't add an external hash (as we use to do with other schemes), we say NONE (meaning "None as external").
4. RSA and Sphincs, although implemented and tested, there are still normalisation changes required to DBs and changes on Artemis queue names, so we currently don't fully support them, due to their large keys (will do in an upcoming PR).
5. I've also made a change required for multi-threaded access to sign and verify (a new Signature object is created per call - we could use a pool of reusable Signature however, to avoid always creating new objects (didn't get into this optimisation trick yet).
6. Had a chat with @patrickkuo, and some small refactoring may be required (mainly making public the SignatureScheme objects (but exposing Strings only)) to also allow X509 cert creation/signing, without hardcoded provider Strings etc. (resolved on my 2nd commit - so this will help Patrick to move some functions inside crypto and have a more elegant abstraction/design (CryptoUtils, TLSUtils, Keystores, X509) and less verbose code).
7. Using Provider's name and exposing Providers can cause unexpected and suspicious behaviour. To circumvent this problem a private map of valid providers has been used (independent on the list registered using Security.addProvider). (resolved)

### CODE REVIEW

**Comment by kchalkias** [comment]
SupportedScheme does only include info values, so no problem to expose it anymore. Schemes are now reachable from the outside world, so we can easily call crypto functions like:
`val keyPair = Crypto.generateKeyPair(Crypto.RSA_SHA256)`
`val keyPair = Crypto.generateKeyPair()` using the default scheme (currently EdDSA)
`val privKey = Crypto.decodePrivateKey(privKey.encoded)` (from PKCS8 encoded)
`val pubKey = Crypto.decodePublicKey(pubKey.encoded)` (from X509 encoded)
'Crypto.ECDSA_SECP256K1_SHA256.schemeCodeName` (autocomplete helps a lot to understand what is supported)

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 1]
```diff
@@ -91,34 +92,33 @@ object Crypto {
...
     )
 
     /** Our default signature scheme if no algorithm is specified (e.g. for key generation). */
-    private val DEFAULT_SIGNATURE_SCHEME = EDDSA_ED25519_SHA512
+    val DEFAULT_SIGNATURE_SCHEME = EDDSA_ED25519_SHA512
 
     /**
      * Supported digital signature schemes.
      * Note: Only the schemes added in this map will be supported (see [Crypto]).
      * Do not forget to add the DEFAULT_SIGNATURE_SCHEME as well.
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: Who is this "do not forget" comment aimed at now it's public?

File: core/src/main/kotlin/net/corda/core/crypto/CryptoUtils.kt [hunk 2]
```diff
@@ -40,34 +37,46 @@ class DummyPublicKey(val s: String) : PublicKey, Comparable<PublicKey> {
...
-fun KeyPair.signWithECDSA(bytesToSign: ByteArray, party: Party): DigitalSignature.LegallyIdentifiable {
-    val sig = signWithECDSA(bytesToSign)
+fun KeyPair.sign(bytesToSign: ByteArray, party: Party): DigitalSignature.LegallyIdentifiable {
+    val sig = sign(bytesToSign)
     val sigKey = when (party.owningKey) { // Quick workaround when we have CompositeKey as Party owningKey.
         is CompositeKey -> throw InvalidKeyException("Signing for parties with CompositeKey not supported.")
         else -> party.owningKey
     }
-    sigKey.verifyWithECDSA(bytesToSign, sig)
+    sigKey.verify(bytesToSign, sig) // TODO: this is not required!
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/CryptoUtils.kt]: OK, so remove it :)

File: core/src/main/kotlin/net/corda/core/crypto/CryptoUtils.kt [hunk 3]
```diff
@@ -82,32 +91,25 @@ fun KeyPair.signWithECDSA(bytesToSign: ByteArray, party: Party): DigitalSignatur
 // we should use another exception (perhaps IllegalArgumentException) for indicating the signature is valid but does
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/CryptoUtils.kt]: Let's either address this TODO now or delete it as unimportant.

**Pull request merged**